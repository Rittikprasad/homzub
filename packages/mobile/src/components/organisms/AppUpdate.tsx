/**
 * Handling Force Update, Maintenance logic using remote config
 * Doc for ref: https://rnfirebase.io/remote-config/usage
 */

import React, { useEffect, useState } from 'react';
import { Linking, StyleSheet, View } from 'react-native';
import remoteConfig from '@react-native-firebase/remote-config';
import { AppModes, ConfigHelper, ConfigModes } from '@homzhub/common/src/utils/ConfigHelper';
import { DeviceUtils } from '@homzhub/common/src/utils/DeviceUtils';
import { Logger } from '@homzhub/common/src/utils/Logger';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { I18nService } from '@homzhub/common/src/services/Localization/i18nextService';
import Rocket from '@homzhub/common/src/assets/images/rocket.svg';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';

// Remote config Parameters
enum ConfigKeys {
  MINIMUM_VERSION = 'minimum_version',
  LATEST_VERSION = 'latest_version',
  IS_UNDER_MAINTENANCE = 'is_under_maintenance',
  MESSAGES = 'messages',
}

interface IMessages {
  update: string;
  maintenance: string;
}

interface IConfig {
  [key: string]: any;
}

interface IOwnState {
  isForceUpdate: boolean;
  isUpdateAvailable: boolean;
  isUnderMaintenance: boolean;
  messages: IMessages;
}

const initialState: IOwnState = {
  isForceUpdate: false,
  isUpdateAvailable: false,
  isUnderMaintenance: false,
  messages: {
    update: '',
    maintenance: '',
  },
};

const AppUpdate = (): React.ReactElement => {
  const [configData, setConfigData] = useState<IConfig[]>([]);
  const [configValues, setConfigValues] = useState<IOwnState>(initialState);
  const [isSheetVisible, setVisibility] = useState(false);

  useEffect(() => {
    remoteConfig()
      .setConfigSettings({
        minimumFetchIntervalMillis: 30000, // Global Caching frequency
      })
      .then(() => {
        // Set Default before fetching config from firebase
        remoteConfig()
          .setDefaults({
            is_under_maintenance: false,
            latest_version: '',
            minimum_version: '',
          })
          .then(() => remoteConfig().fetchAndActivate()) // Fetch config and Activate
          .then(() => {
            const config = remoteConfig().getAll(); // Get All config list

            if (config) {
              const formattedData = Object.entries(config).map((item) => {
                const [key, entry] = item as any;
                return {
                  [key]: JSON.parse(entry.asString()),
                };
              });
              setConfigData(formattedData);
            }
          })
          .then(() => remoteConfig().fetch()); // Fetching data for next process
      })

      .catch((e) => {
        Logger.warn(e.message);
      });
  }, []);

  useEffect(() => {
    const updatedData = { ...configValues };
    const configMode = ConfigHelper.getConfigMode();
    const appMode = (configMode === ConfigModes.PROD ? AppModes.RELEASE : AppModes.DEBUG).toLowerCase();

    // Structuring config and maintaining local state to handle UI
    configData.forEach((item) => {
      if (Object.keys(item)[0] === ConfigKeys.MINIMUM_VERSION) {
        updatedData.isForceUpdate = item[ConfigKeys.MINIMUM_VERSION][appMode] > DeviceUtils.appVersion;
      } else if (Object.keys(item)[0] === ConfigKeys.LATEST_VERSION) {
        updatedData.isUpdateAvailable = item[ConfigKeys.LATEST_VERSION][appMode] > DeviceUtils.appVersion;
      } else if (Object.keys(item)[0] === ConfigKeys.IS_UNDER_MAINTENANCE) {
        updatedData.isUnderMaintenance = item[ConfigKeys.IS_UNDER_MAINTENANCE][appMode];
      } else {
        updatedData.messages = item[ConfigKeys.MESSAGES];
      }
    });

    const isVisible = updatedData.isUpdateAvailable || updatedData.isUnderMaintenance || updatedData.isForceUpdate;

    setVisibility(isVisible);
    setConfigValues(updatedData);
  }, [configData]);

  const onUpdateNow = async (): Promise<void> => {
    const { getAppleStoreUrl, getGooglePlayStoreUrl } = ConfigHelper;
    await Linking.openURL(PlatformUtils.isIOS() ? getAppleStoreUrl() : getGooglePlayStoreUrl());
  };

  const { isForceUpdate, messages, isUnderMaintenance, isUpdateAvailable } = configValues;
  const isAlwaysVisible = isForceUpdate || isUnderMaintenance;
  return (
    <BottomSheet
      visible={isSheetVisible}
      isShadowView
      headerTitle={isUpdateAvailable ? I18nService.t('common:timeToUpdate') : I18nService.t('common:underMaintenance')}
      sheetHeight={isUpdateAvailable ? 420 : 300}
      isCloseOnDrag={!isAlwaysVisible}
      isAlwaysVisible={isAlwaysVisible}
      onCloseSheet={(): void => setVisibility(false)}
    >
      <View style={styles.container}>
        <Rocket style={styles.iconStyle} />
        <Label type="large" style={styles.message}>
          {isUpdateAvailable ? messages?.update : messages?.maintenance}
        </Label>
        {!isUnderMaintenance && (
          <Button
            type="primary"
            title={I18nService.t('common:updateNow')}
            containerStyle={styles.updateButton}
            onPress={onUpdateNow}
          />
        )}
        {!isAlwaysVisible && (
          <Button
            type="secondaryOutline"
            title={I18nService.t('common:doLater')}
            containerStyle={styles.secondaryButton}
            textStyle={styles.buttonTitle}
            onPress={(): void => setVisibility(false)}
          />
        )}
      </View>
    </BottomSheet>
  );
};

export default AppUpdate;

const styles = StyleSheet.create({
  container: {
    padding: 32,
    flex: 1,
  },
  iconStyle: {
    alignSelf: 'center',
  },
  message: {
    textAlign: 'center',
    marginVertical: 16,
    color: theme.colors.darkTint3,
  },
  updateButton: {
    flex: 0,
    height: 46,
    marginVertical: 10,
  },
  secondaryButton: {
    height: 46,
    flex: 0,
  },
  buttonTitle: {
    color: theme.colors.darkTint3,
  },
});
