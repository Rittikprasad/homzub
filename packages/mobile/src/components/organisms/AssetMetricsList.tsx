import React, { useState, useCallback } from 'react';
import { StyleProp, StyleSheet, TouchableOpacity, View, ViewStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { useRoute } from '@react-navigation/native';
import { useNavigation } from '@react-navigation/core';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import Logo from '@homzhub/common/src/assets/images/homzhubDashboard.svg';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { PaginationComponent } from '@homzhub/mobile/src/components/atoms/PaginationComponent';
import { SnapCarousel } from '@homzhub/mobile/src/components/atoms/Carousel';
import { AssetMetrics } from '@homzhub/mobile/src/components/molecules/AssetMetrics';

export interface IMetricsData {
  name: string;
  count: string | number;
  label?: string;
  id?: number;
  isCurrency?: boolean;
  colorCode: string;
  code?: string;
}

interface IProps {
  data: IMetricsData[];
  title?: string;
  subscription?: string;
  selectedAssetType?: string;
  numOfElements?: number;
  isSubTextRequired?: boolean;
  onPlusIconClicked?: () => void;
  onMetricsClicked?: (name: string) => void;
  containerStyle?: StyleProp<ViewStyle>;
  subTitleText?: string;
  headerIcon?: string;
  showBackIcon?: boolean;
}

const COMPONENT_PADDING = 12;
const SLIDER_WIDTH = theme.viewport.width - theme.layout.screenPadding * 2;
const AssetMetricsList = (props: IProps): React.ReactElement => {
  const {
    title = 0,
    data,
    selectedAssetType,
    onPlusIconClicked,
    containerStyle,
    onMetricsClicked,
    numOfElements = 3,
    isSubTextRequired = true,
    subTitleText,
    headerIcon,
    showBackIcon = false,
  } = props;

  // HOOKS
  const { t } = useTranslation();
  const { goBack, setParams } = useNavigation();
  const { params } = useRoute();
  const [activeIndex, setActiveIndex] = useState(0);
  // HELPERS
  const bubblePlusIcon = useCallback((): void => {
    if (onPlusIconClicked) {
      onPlusIconClicked();
    }
  }, [onPlusIconClicked]);

  const splitData = useCallback((): IMetricsData[][] => {
    const newArr = [];

    for (let i = 0; i < data.length; i += numOfElements) {
      newArr.push(data.slice(i, i + numOfElements));
    }

    return newArr;
  }, [data, numOfElements]);

  const onGoBack = (): void => {
    goBack();
    // @ts-ignore
    if (params && params?.isFromNavigation) {
      setParams({ isFromNavigation: false });
    }
  };
  // HELPERS END

  const renderItem = useCallback(
    (items: IMetricsData[]): React.ReactElement => {
      return (
        <View style={styles.sliderRow}>
          {items.map((item: IMetricsData) => {
            const cardStyle = {
              minWidth: (SLIDER_WIDTH - COMPONENT_PADDING * (numOfElements + 1)) / numOfElements,
            };
            // TODO: (Shikha) - Connect with BE to handle all logic with CODE
            const handlePress = (): void => onMetricsClicked && onMetricsClicked(item.code || item.name);
            return (
              <AssetMetrics
                key={`${item.label ?? item.name}`}
                header={item.label ?? item.name}
                value={item.count}
                isCurrency={item.isCurrency ?? false}
                cardStyle={cardStyle}
                colorCode={item.colorCode}
                onPressMetrics={handlePress}
                selectedAssetType={selectedAssetType}
                // Todo : Check in multiple devices
                maxLength={items.length === 2 ? 15 : undefined}
              />
            );
          })}
        </View>
      );
    },
    [onMetricsClicked, numOfElements, selectedAssetType]
  );

  return (
    <View style={[styles.container, containerStyle]}>
      <View style={[styles.header, !isSubTextRequired && styles.financialView]}>
        {showBackIcon && (
          <TouchableOpacity onPress={onGoBack} style={styles.backIcon}>
            <Icon name={icons.leftArrow} size={25} color={theme.colors.blue} />
          </TouchableOpacity>
        )}
        {headerIcon ? (
          <View>
            <Icon name={headerIcon} size={30} style={styles.icon} color={theme.colors.blue} />
            <View style={[styles.headerIcon, styles.logo]} />
          </View>
        ) : (
          <Logo style={styles.logo} width={50} height={50} />
        )}
        <View style={styles.headerCenter}>
          <Text type="regular" textType="bold" style={styles.assetCount}>
            {title}
          </Text>
          <View style={styles.propertiesRow}>
            {isSubTextRequired && (
              <Label type="large" textType="semiBold" style={styles.propertyText}>
                {subTitleText || t('common:properties')}
              </Label>
            )}
          </View>
        </View>
        {onPlusIconClicked && (
          <TouchableOpacity onPress={bubblePlusIcon} style={styles.plusIconContainer}>
            <Icon name={icons.circularPlus} size={32} color={theme.colors.primaryColor} testID="icnPlus" />
          </TouchableOpacity>
        )}
      </View>
      <SnapCarousel
        carouselData={splitData()}
        carouselItem={renderItem}
        activeIndex={activeIndex}
        onSnapToItem={setActiveIndex}
        sliderWidth={SLIDER_WIDTH}
        itemWidth={SLIDER_WIDTH}
      />
      <PaginationComponent
        dotsLength={splitData().length}
        activeSlide={activeIndex}
        containerStyle={styles.paginationContainer}
        activeDotStyle={[styles.dot, styles.activeDot]}
        inactiveDotStyle={[styles.dot, styles.inactiveDot]}
      />
    </View>
  );
};

const memoizedComponent = React.memo(AssetMetricsList);
export { memoizedComponent as AssetMetricsList };

const styles = StyleSheet.create({
  container: {
    paddingVertical: 12,
    borderRadius: 4,
    backgroundColor: theme.colors.white,
  },
  header: {
    marginHorizontal: 12,
    flexDirection: 'row',
  },
  headerCenter: {
    flex: 1,
  },
  propertiesRow: {
    flexDirection: 'row',
  },
  sliderRow: {
    flexDirection: 'row',
  },
  paginationContainer: {
    paddingVertical: 0,
    marginTop: 12,
  },
  assetCount: {
    color: theme.colors.darkTint1,
  },
  propertyText: {
    color: theme.colors.darkTint4,
  },
  logo: {
    marginEnd: 12,
  },
  dot: {
    width: 8.5,
    height: 8.5,
  },
  activeDot: {
    borderWidth: 1.5,
  },
  inactiveDot: {
    backgroundColor: theme.colors.disabled,
    borderWidth: 0,
  },
  financialView: {
    alignItems: 'center',
  },
  headerIcon: {
    backgroundColor: theme.colors.blue,
    width: 50,
    height: 50,
    opacity: 0.1,
    borderRadius: 30,
  },
  icon: {
    position: 'absolute',
    left: 10,
    top: 10,
  },
  plusIconContainer: {
    justifyContent: 'center',
    padding: 5,
  },
  backIcon: {
    alignSelf: 'center',
    marginRight: 10,
  },
});
