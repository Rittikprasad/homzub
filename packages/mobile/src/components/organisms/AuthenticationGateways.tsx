import React from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { WithTranslation, withTranslation } from 'react-i18next';
import { StackNavigationProp } from '@react-navigation/stack';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import { SocialAuthUtils } from '@homzhub/common/src/utils/SocialAuthUtils';
import { AuthService } from '@homzhub/mobile/src/services/AuthService';
import { ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import Apple from '@homzhub/common/src/assets/images/apple.svg';
import Google from '@homzhub/common/src/assets/images/google.svg';
import Facebook from '@homzhub/common/src/assets/images/facebook.svg';
import LinkedIn from '@homzhub/common/src/assets/images/linkedin.svg';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import CommonAuthenticationHoc from '@homzhub/common/src/components/organisms/CommonAuthenticationHOC';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { SocialAuthProvider } from '@homzhub/common/src/domain/models/SocialAuthProvider';
import { ISocialUserData, SocialAuthKeys } from '@homzhub/common/src/constants/SocialAuthProviders';
import { IUserTokens } from '@homzhub/common/src/services/storage/StorageService';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IDispatchProps {
  loginSuccess: (data: IUserTokens) => void;
}

interface IAuthenticationGatewayProps {
  isFromLogin: boolean;
  onEmailLogin?: () => void;
  testID?: string;
  navigation?: StackNavigationProp<AuthStackParamList, ScreensKeys.SignUp | ScreensKeys.Login>;
  toggleLoading?: (isLoading: boolean) => void;
  onSuccessCallback?: () => void;
  isSocialAllowed?: boolean;
}

type Props = IAuthenticationGatewayProps & WithTranslation & IDispatchProps;

interface IOwnState {
  authProviders: SocialAuthProvider[];
}

class AuthenticationGateways extends React.PureComponent<Props, IOwnState> {
  public state = {
    authProviders: [],
  };

  public componentDidMount(): void {
    SocialAuthUtils.fetchSocialMedia((response) => {
      if (PlatformUtils.isIOS()) {
        response.unshift(
          ObjectMapper.deserialize(SocialAuthProvider, {
            provider: SocialAuthKeys.Apple,
            description: 'Apple',
            visible: PlatformUtils.isIOS(),
          })
        );
      }

      this.setState({ authProviders: response });
    });
  }

  public render(): React.ReactNode {
    const { onEmailLogin, isFromLogin, isSocialAllowed = true } = this.props;

    return (
      <CommonAuthenticationHoc isFromLogin={isFromLogin} onEmailLogin={onEmailLogin}>
        {!PlatformUtils.isIOS() && isSocialAllowed && this.renderSocialAuths()}
      </CommonAuthenticationHoc>
    );
  }

  private renderSocialAuths = (): React.ReactNode => {
    const { authProviders } = this.state;

    return authProviders.map((auth: SocialAuthProvider) => {
      const initiateSocialLogin = async (): Promise<void> => {
        await this.initiateSocialAuthentication(auth.provider);
      };

      return (
        <>
          {auth.visible && (
            <TouchableOpacity style={styles.alignToCenter} onPress={initiateSocialLogin} key={auth.provider}>
              {this.getImageUrlForLib(auth.provider)}
              <Label type="regular" textType="regular" style={styles.iconTextStyle}>
                {auth.description}
              </Label>
            </TouchableOpacity>
          )}
        </>
      );
    });
  };

  private onSocialAuthSuccess = async (userData: ISocialUserData): Promise<void> => {
    const { loginSuccess, navigation, isFromLogin, onSuccessCallback } = this.props;

    if (navigation) {
      await SocialAuthUtils.onSocialAuthSuccess(
        userData,
        () => {
          navigation.navigate(ScreensKeys.MobileVerification, {
            isFromLogin,
            userData,
          });
        },
        (response) => {
          loginSuccess(response);
          if (onSuccessCallback) {
            onSuccessCallback();
          }
        }
      );
    }
  };

  private initiateSocialAuthentication = async (key: string): Promise<void> => {
    const { isFromLogin, navigation } = this.props;
    this.setLoading(true);
    if (key === SocialAuthKeys.Google) {
      await AuthService.signInWithGoogle(this.onSocialAuthSuccess);
      this.setLoading(false);
      return;
    }

    if (key === SocialAuthKeys.Apple) {
      await AuthService.signInWithApple(this.onSocialAuthSuccess);
      this.setLoading(false);
      return;
    }

    if (key === SocialAuthKeys.Facebook) {
      await AuthService.signInWithFacebook(this.onSocialAuthSuccess, isFromLogin && navigation);
      this.setLoading(false);
    }
  };

  private setLoading = (isLoading: boolean): void => {
    const { toggleLoading } = this.props;
    if (toggleLoading) {
      toggleLoading(isLoading);
    }
  };

  private getImageUrlForLib = (key: string): React.ReactNode => {
    switch (key) {
      case SocialAuthKeys.Facebook:
        return <Facebook />;
      case SocialAuthKeys.Google:
        return <Google />;
      case SocialAuthKeys.LinkedIn:
        return <LinkedIn />;
      case SocialAuthKeys.Apple:
        return <Apple />;
      default:
        return null;
    }
  };
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { loginSuccess } = UserActions;
  return bindActionCreators(
    {
      loginSuccess,
    },
    dispatch
  );
};

const HOC = connect(
  null,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.auth)(AuthenticationGateways));
export { HOC as AuthenticationGateways };

const styles = StyleSheet.create({
  iconTextStyle: {
    marginTop: 4,
  },
  alignToCenter: {
    alignItems: 'center',
  },
});
