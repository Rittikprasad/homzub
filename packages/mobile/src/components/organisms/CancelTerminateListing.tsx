import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Formik, FormikProps, FormikValues } from 'formik';
import { WithTranslation, withTranslation } from 'react-i18next';
import * as yup from 'yup';
import { DateFormats, DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import { Label } from '@homzhub/common/src/components/atoms/Text';
import { TextArea } from '@homzhub/common/src/components/atoms/TextArea';
import { FormButton } from '@homzhub/common/src/components/molecules/FormButton';
import { FormCalendar } from '@homzhub/common/src/components/molecules/FormCalendar';
import { FormDropdown, IDropdownOption } from '@homzhub/common/src/components/molecules/FormDropdown';

export interface ISubmitPayload extends IFormData {
  isTerminate: boolean;
  description: string;
}

export interface IFormData {
  reasonId: number;
  terminationDate: string;
}

interface IScreenState {
  formData: IFormData;
  description: string;
  isTouched: boolean;
}

interface IProps {
  isTerminate?: boolean;
  leaseEndDate?: string;
  onFormEdit: () => void;
  reasonData: IDropdownOption[];
  onSubmit: (payload: ISubmitPayload) => void;
}

type Props = IProps & WithTranslation;

class CancelTerminateListing extends Component<Props, IScreenState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      formData: {
        reasonId: 0,
        terminationDate: '',
      },
      description: '',
      isTouched: false,
    };
  }

  public render(): React.ReactNode {
    const { formData, description, isTouched } = this.state;
    const { t, isTerminate = false, onFormEdit, reasonData, leaseEndDate } = this.props;

    return (
      <View style={styles.form}>
        <Formik
          onSubmit={this.handleSubmit}
          initialValues={{ ...formData }}
          validate={FormUtils.validate(this.formSchema)}
        >
          {(formProps: FormikProps<FormikValues>): React.ReactNode => {
            const { reasonId, terminationDate } = formProps.values;
            const isButtonEnable = isTerminate ? reasonId > 0 && !!terminationDate : reasonId > 0;

            if (!isTouched && formProps.dirty) {
              this.setState({ isTouched: true });
              onFormEdit();
            }

            return (
              <>
                <FormDropdown
                  name="reasonId"
                  placeholder={t('common:reasonExample')}
                  label={t('common:reason')}
                  isMandatory
                  options={reasonData}
                  formProps={formProps}
                  dropdownContainerStyle={styles.dropdownStyle}
                />
                {isTerminate && (
                  <FormCalendar
                    formProps={formProps}
                    isCurrentDateEnable
                    maxDate={leaseEndDate && DateUtils.getUtcFormatted(leaseEndDate, DateFormats.DD_MM_YYYY)}
                    name="terminationDate"
                    textType="label"
                    label={t('property:terminationDate')}
                    calendarTitle={t('property:selectMoveOutDate')}
                    placeHolder={t('property:selectMoveOutDate')}
                    isMandatory
                  />
                )}
                <TextArea
                  value={description}
                  placeholder={t('common:typeDescriptionHere')}
                  label={t('assetDescription:description')}
                  wordCountLimit={500}
                  containerStyle={styles.textArea}
                  onMessageChange={this.handleDescription}
                />
                <Label type="large" style={styles.message}>
                  {t('common:actionNotDone')}
                </Label>
                <FormButton
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  containerStyle={{ backgroundColor: isButtonEnable ? theme.colors.error : theme.colors.disabled }}
                  formProps={formProps}
                  disabled={!isButtonEnable}
                  type="primary"
                  title={isTerminate ? t('common:sendNotice') : t('property:deleteListing')}
                />
              </>
            );
          }}
        </Formik>
      </View>
    );
  }

  private handleDescription = (value: string): void => {
    this.setState({ description: value });
  };

  private formSchema = (): yup.ObjectSchema<IFormData> => {
    const { t } = this.props;

    return yup.object().shape({
      terminationDate: yup.string().when('isTerminate', {
        is: true,
        then: yup.string().required(t('moreProfile:fieldRequiredError')),
      }),
      reasonId: yup.number().required(t('moreProfile:fieldRequiredError')),
    });
  };

  private handleSubmit = (values: IFormData): void => {
    const { onSubmit, isTerminate } = this.props;
    const { description } = this.state;
    onSubmit({ ...values, isTerminate: isTerminate ?? false, description });
  };
}

export default withTranslation()(CancelTerminateListing);

const styles = StyleSheet.create({
  form: {
    paddingVertical: 14,
  },
  dropdownStyle: {
    paddingVertical: 12,
  },
  textArea: {
    marginTop: 10,
    marginBottom: 50,
  },
  message: {
    alignSelf: 'center',
    color: theme.colors.darkTint5,
    marginVertical: 12,
  },
});
