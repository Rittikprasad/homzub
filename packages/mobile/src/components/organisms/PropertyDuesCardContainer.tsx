import React, { ReactElement, ReactNode } from 'react';
import { StyleSheet, View } from 'react-native';
import { useTranslation } from 'react-i18next';
import { TFunction } from 'i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Divider } from '@homzhub/common/src/components/atoms/Divider';
import { Image } from '@homzhub/common/src/components/atoms/Image';
import { PricePerUnit } from '@homzhub/common/src/components/atoms/PricePerUnit';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';

interface IPropertyDuesData {
  propertyName: string;
  address: string;
  dueCategory: string;
  price: number;
  currency_symbol: string;
}

interface IProps {
  propertyDues: IPropertyDuesData[];
  totalDue: number;
  currency: string;
}

export const PropertyDuesCardContainer = ({ ...props }: IProps): ReactElement => {
  const { totalDue, currency, propertyDues } = props;
  const { t } = useTranslation();

  return (
    <View style={styles.containerStyle}>
      <View style={styles.heading}>
        <View style={styles.duesText}>
          <Icon style={styles.walletIconStyle} name={icons.wallet} size={22} />
          <Text type="small" textType="semiBold">
            {t('assetDashboard:dues')}
          </Text>
        </View>
        <PricePerUnit
          textSizeType="small"
          textStyle={styles.dueStyle}
          priceTransformation={false}
          currency={currency}
          price={totalDue}
        />
      </View>
      <Divider />
      <View style={styles.propertyDuesContainer}>{renderPropertyDues(propertyDues, t)}</View>
    </View>
  );
};

const renderPropertyDues = (propertyDues: IPropertyDuesData[], t: TFunction): ReactNode => {
  return propertyDues.map((property, index) => {
    const { propertyName, address, dueCategory, price, currency_symbol } = property;

    return (
      <View key={`${property.propertyName}${index}`}>
        <View style={styles.propertyName}>
          <Image style={styles.flag} source={{ uri: '' }} />
          <Text type="small" textType="semiBold">
            {propertyName}
          </Text>
        </View>
        <Label type="large">{address}</Label>
        <View style={styles.dueContainer}>
          <View>
            <Label type="large">{dueCategory}</Label>
            <PricePerUnit textSizeType="small" priceTransformation={false} currency={currency_symbol} price={price} />
          </View>
          <Button
            textType="text"
            textSize="small"
            titleStyle={styles.payNowTitleStyles}
            containerStyle={styles.payNowBtnStyles}
            type="primary"
            title={t('assetFinancial:payNow')}
          />
        </View>
        {index !== propertyDues.length - 1 ? <Divider containerStyles={styles.dividerStyles} /> : null}
      </View>
    );
  });
};

const styles = StyleSheet.create({
  containerStyle: {
    marginTop: 16,
    backgroundColor: theme.colors.white,
    borderRadius: 4,
  },
  duesText: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  dueStyle: {
    color: theme.colors.danger,
  },
  propertyDuesContainer: {
    padding: 16,
  },
  dividerStyles: {
    marginVertical: 24,
  },
  heading: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 16,
  },
  flag: {
    marginRight: 6,
  },
  propertyName: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  dueContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginTop: 16,
  },
  payNowBtnStyles: {
    flex: 0,
  },
  payNowTitleStyles: {
    marginVertical: 6,
    marginHorizontal: 24,
  },
  walletIconStyle: {
    marginRight: 10,
  },
});
