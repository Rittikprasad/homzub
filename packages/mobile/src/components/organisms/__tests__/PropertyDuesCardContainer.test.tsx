import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyDuesCardContainer } from '@homzhub/mobile/src/components/organisms/PropertyDuesCardContainer';
import { propertyDues } from '@homzhub/common/src/mocks/FinancialsTabMockData';

let props: any;
let wrapper: ShallowWrapper;

describe('PropertyDuesCardContainer', () => {
  const createTestProps = (testProps: any): object => ({
    propertyDues: propertyDues.details,
    totalDue: 1200,
    currency: 'INR',
    ...testProps,
  });

  props = createTestProps({});

  it('should match snapshot', () => {
    wrapper = shallow(<PropertyDuesCardContainer {...props} t={(key: string): string => key} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
