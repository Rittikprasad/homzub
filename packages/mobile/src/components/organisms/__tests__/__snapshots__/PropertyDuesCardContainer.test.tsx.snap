// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyDuesCardContainer should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 4,
      "marginTop": 16,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Icon
        allowFontScaling={false}
        name="wallet"
        size={22}
        style={
          Object {
            "marginRight": 10,
          }
        }
      />
      <Text
        textType="semiBold"
        type="small"
      >
        assetDashboard:dues
      </Text>
    </View>
    <Memo(PricePerUnit)
      currency="INR"
      price={1200}
      priceTransformation={false}
      textSizeType="small"
      textStyle={
        Object {
          "color": "#D60D31",
        }
      }
    />
  </View>
  <Divider />
  <View
    style={
      Object {
        "padding": 16,
      }
    }
  >
    <View
      key="Eaton Garth Manor0"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "marginRight": 6,
            }
          }
        />
        <Text
          textType="semiBold"
          type="small"
        >
          Eaton Garth Manor
        </Text>
      </View>
      <Label
        type="large"
      >
        Sindhi Society, Chembur, Mumbai- 400071
      </Label>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 16,
          }
        }
      >
        <View>
          <Label
            type="large"
          >
            Plumbing fees
          </Label>
          <Memo(PricePerUnit)
            currency="USD"
            price={20000}
            priceTransformation={false}
            textSizeType="small"
          />
        </View>
        <Button
          containerStyle={
            Object {
              "flex": 0,
            }
          }
          textSize="small"
          textType="text"
          title="assetFinancial:payNow"
          titleStyle={
            Object {
              "marginHorizontal": 24,
              "marginVertical": 6,
            }
          }
          type="primary"
        />
      </View>
      <Divider
        containerStyles={
          Object {
            "marginVertical": 24,
          }
        }
      />
    </View>
    <View
      key="Eaton Garth Manor1"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "marginRight": 6,
            }
          }
        />
        <Text
          textType="semiBold"
          type="small"
        >
          Eaton Garth Manor
        </Text>
      </View>
      <Label
        type="large"
      >
        Sindhi Society, Chembur, Mumbai- 400071
      </Label>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 16,
          }
        }
      >
        <View>
          <Label
            type="large"
          >
            Plumbing fees
          </Label>
          <Memo(PricePerUnit)
            currency="USD"
            price={30000}
            priceTransformation={false}
            textSizeType="small"
          />
        </View>
        <Button
          containerStyle={
            Object {
              "flex": 0,
            }
          }
          textSize="small"
          textType="text"
          title="assetFinancial:payNow"
          titleStyle={
            Object {
              "marginHorizontal": 24,
              "marginVertical": 6,
            }
          }
          type="primary"
        />
      </View>
      <Divider
        containerStyles={
          Object {
            "marginVertical": 24,
          }
        }
      />
    </View>
    <View
      key="Eaton Garth Manor2"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "marginRight": 6,
            }
          }
        />
        <Text
          textType="semiBold"
          type="small"
        >
          Eaton Garth Manor
        </Text>
      </View>
      <Label
        type="large"
      >
        Sindhi Society, Chembur, Mumbai- 400071
      </Label>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 16,
          }
        }
      >
        <View>
          <Label
            type="large"
          >
            Plumbing fees
          </Label>
          <Memo(PricePerUnit)
            currency="USD"
            price={20000}
            priceTransformation={false}
            textSizeType="small"
          />
        </View>
        <Button
          containerStyle={
            Object {
              "flex": 0,
            }
          }
          textSize="small"
          textType="text"
          title="assetFinancial:payNow"
          titleStyle={
            Object {
              "marginHorizontal": 24,
              "marginVertical": 6,
            }
          }
          type="primary"
        />
      </View>
    </View>
  </View>
</View>
`;
